p/api/fever.php

```diff
@@ -0,0 +1,649 @@
+		Minz_Session::_param('currentUser');
+		$feverKey = empty($_POST['api_key']) ? '' : substr(trim($_POST['api_key']), 0, 128);
+		if (ctype_xdigit($feverKey)) {
+			$feverKey = strtolower($feverKey);
+			$username = @file_get_contents(DATA_PATH . '/fever/.key-' . sha1(FreshRSS_Context::$system_conf->salt) . '-' . $feverKey . '.txt', false);
+			if ($username != false) {
+				$username = trim($username);
+				//Check that it is not an old forgotten key
+				$reverseKey = @file_get_contents(DATA_PATH . '/fever/.user-' . sha1(FreshRSS_Context::$system_conf->salt) . '-' . $username . '.txt', false);
+				$reverseKey = $reverseKey == false ? '' : trim($reverseKey);
+				if ($feverKey === $reverseKey && FreshRSS_user_Controller::checkUsername($username)) {
+					FreshRSS_Context::$user_conf = get_user_configuration($username);
+					if (FreshRSS_Context::$user_conf != null) {
+						Minz_Session::_param('currentUser', $username);
+					}
+				}
+			}
+		}
+	}
+
+	/**
+	 * @return bool
+	 */
+	public function isAuthenticatedApiUser()
+	{
+		if (FreshRSS_Context::$user_conf !== null) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * @return FreshRSS_FeedDAO
+	 */
+	protected function getDaoForFeeds()
+	{
+		return new FreshRSS_FeedDAO();
+	}
+
+	/**
+	 * @return FreshRSS_CategoryDAO
+	 */
+	protected function getDaoForCategories()
+	{
+		return new FreshRSS_CategoryDAO();
+	}
+
+	/**
+	 * @return FeverAPI_EntryDAO
+	 */
+	protected function getDaoForEntries()
+	{
+		return new FeverAPI_EntryDAO();
+	}
+
+	/**
+	 * this does all the processing, since the fever api does not have a specific variable that specifies the operation
+	 */
+	public function process()
+	{
+		$response_arr = array();
+
+		if (!$this->isAuthenticatedApiUser()) {
+			throw new Exception('No user given or user is not allowed to access API');
+		}
+
+		if (isset($_REQUEST["groups"])) {
+			$response_arr["groups"] = $this->getGroups();
+			$response_arr["feeds_groups"] = $this->getFeedsGroup();
+		}
+
+		if (isset($_REQUEST["feeds"])) {
+			$response_arr["feeds"] = $this->getFeeds();
+			$response_arr["feeds_groups"] = $this->getFeedsGroup();
+		}
+
+		if (isset($_REQUEST["favicons"])) {
+			$response_arr["favicons"] = $this->getFavicons();
+		}
+
+		if (isset($_REQUEST["items"])) {
+			$response_arr["total_items"] = $this->getTotalItems();
+			$response_arr["items"] = $this->getItems();
+		}
+
+		if (isset($_REQUEST["links"])) {
+			$response_arr["links"] = $this->getLinks();
+		}
+
+		if (isset($_REQUEST["unread_item_ids"])) {
+			$response_arr["unread_item_ids"] = $this->getUnreadItemIds();
+		}
+
+		if (isset($_REQUEST["saved_item_ids"])) {
+			$response_arr["saved_item_ids"] = $this->getSavedItemIds();
+		}
+
+		if (isset($_REQUEST["mark"], $_REQUEST["as"], $_REQUEST["id"]) && is_numeric($_REQUEST["id"])) {
+			$method_name = "set" . ucfirst($_REQUEST["mark"]) . "As" . ucfirst($_REQUEST["as"]);
+			if (method_exists($this, $method_name)) {
```

Comment:
I am a bit worried about calling methods based on user input. I would rather like to test (e.g. with a switch) against a white-list of acceptable methods, instead of solely `method_exists()`.